docs/src/gradient_accumulation.md

```diff
@@ -0,0 +1,134 @@
+
+### Maybe-mutating accumulation (`add!!`)
+We can note that in the above that neither `ā` nor `b̄` are ever used again after accumulating to get `X̄`.
+Further `Array`s are mutable.
+That means we could over-write either `ā` or `b̄` and use the result as `X̄`:
+```julia
+ā .+= b̄
+X̄ = ā
+```
+This cuts our allocations down to 2, just `ā` and `b̄`.
+
+However, we have a bit of a problem that not all types are mutable, so this pattern is hard to apply in general.
+To deal with that ChainRulesCore provides [`add!!`](@ref).
+Per the [BangBang.jl](https://github.com/JuliaFolds/BangBang.jl) convention, this is a maybe mutating addition.
+It may mutating its first argument (it it is mutable), but it will definitely return the correct result.
+We would write using that as `X̄ = add!!(ā, b̄)`: which would in this case give us just 2 allocations.
+AD systems can generate `add!!` instead of `+` when accumulating gradient to take advantage of this.
+
+### Inplaceable Thunks (`InplaceableThunks`) avoid allocating values in the first place.
+We got down to two allocations from using [`add!!`](@ref), but can we do better?
+We can think of having differential type which acts on a partially accumulated result, to mutate it to contain its current value plus the partial derivative being accumulated.
+Rather than having a actual computed value, we can just have a thing that will act on a value to perform the addition.
+This is a bit of a funny idea, lets illustrate it with our example.
+
+`b̄` is the partial for `X[2]`.
+It's value can be computed by:
+```julia
+b̄ = zeros(size(X))
+b̄[2] = ȳ  # the scalar sensitivity of the `mysum` output
+```
+`b̄` is a matrix entirely of zeros, except for at the index `2`, where it is set to the output sensitivity `ȳ`.
+`ā` is similar, except with the non-zero at index `1`.
+
+What is the action of `b̄` upon `ā`, to get the same result as `X̄ = add!!(ā, b̄)` (or `X̄ = ā + b̄` for that matter)?
+It is:
+```julia
+function b̄_add!(ā)
+    ā[2] += ȳ
+    return ā
+end
+```
+We don't need to worry about all those zeros since `x + 0 == x`.
+
+[`InplaceableThunk`](@ref) is the type we have to represent derivatives as gradient accumulating actions.
+We must note that to do this we do need a value form of `ā` for `b̄` to act upon.
+For this reason every inplaceable thunk has both a `val` field holding the value representation, and a `add!` field holding the action representation.
+The `val` field use a plain [`Thunk`](@ref) to avoid the computation (and thus allocation) for if it is unused.
+
+!!! note "Do we need both representations?"
+    Right now every [`InplaceableThunk`](@ref) has two fields that need to be specified.
+    The value form (represented as a the `Thunk`(@ref) typed field), and the action form (represented as the `add!` field).
+    It is possible in a future version of ChainRulesCore.jl we will work out a clever way to find the zero differential for arbitrary primal values.
```

Comment:
Possibly a digression from this point on in the note? Maybe we should write this up once it is ready, rather than discuss potential plans here?

Reply:
good question.
I think we can be forgiven in a sidenote.
And i think understandign that this is a limitation helped understand why we do this.
Rather than thinking that it was intrinsically true that you need both value and action form.